pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_controller.go

```diff
@@ -46,48 +50,58 @@ func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, req kube_ctrl.Reque
 		return kube_ctrl.Result{}, err
 	}
 
-	// TODO check that gateways exist etc
-	// TODO set status on gapi resources
-
 	coreName := util_k8s.K8sNamespacedNameToCoreName(httpRoute.Name, httpRoute.Namespace)
 	mesh := k8s_util.MeshFor(httpRoute)
 
 	resource := core_mesh.NewGatewayRouteResource()
 
-	if err := manager.Upsert(r.ResourceManager, model.ResourceKey{Mesh: mesh, Name: coreName}, resource, func(resource model.Resource) error {
-		spec, err := r.gapiToKumaRoute(ctx, mesh, httpRoute.Namespace, httpRoute)
-		if err != nil {
-			return errors.Wrap(err, "error generating GatewayRoute")
-		}
+	specs, status, err := r.gapiToKumaRoutes(ctx, mesh, httpRoute)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "error generating GatewayRoute")
+	}
 
-		return resource.SetSpec(spec)
-	}); err != nil {
-		return kube_ctrl.Result{}, errors.Wrap(err, "could not upsert GatewayRoute")
+	for _, spec := range specs {
+		if err := manager.Upsert(r.ResourceManager, model.ResourceKey{Mesh: mesh, Name: coreName}, resource, func(resource model.Resource) error {
+			return resource.SetSpec(spec)
+		}); err != nil {
+			return kube_ctrl.Result{}, errors.Wrap(err, "could not upsert GatewayRoute")
+		}
 	}
 
-	var err error
+	httpRoute.Status = r.mergeStatus(ctx, httpRoute, status)
 
-	resource.Spec, err = r.gapiToKumaRoute(ctx, mesh, httpRoute.Namespace, httpRoute)
-	if err != nil {
-		return kube_ctrl.Result{}, err
+	if err := r.Client.Status().Update(ctx, httpRoute); err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to update HTTPRoute status")
 	}
 
-	err = r.ResourceManager.Update(ctx, resource, store.ModifiedAt(core.Now()))
-
-	return kube_ctrl.Result{}, errors.Wrap(err, "could not update GatewayRoute resource")
+	return kube_ctrl.Result{}, nil
 }
 
-func (r *HTTPRouteReconciler) gapiToKumaRoute(ctx context.Context, mesh string, namespace string, route *gatewayapi.HTTPRoute) (*mesh_proto.GatewayRoute, error) {
-	var selectors []*mesh_proto.Selector
+// gapiToKumaRoutes returns some number of GatewayRoutes that should be created
+// for this HTTPRoute along with any statuses to be set on the HTTPRoute.
+// Only unexpected errors are returned as error.
+func (r *HTTPRouteReconciler) gapiToKumaRoutes(
+	ctx context.Context, mesh string, route *gatewayapi.HTTPRoute,
+) ([]*mesh_proto.GatewayRoute, ParentStatuses, error) {
+	refRoutes := map[gatewayapi.ParentRef]*mesh_proto.GatewayRoute{}
+
+	// Convert GAPI parent refs into some number of GatewayRoutes with Kuma tag
+	// matchers
+	for i, ref := range route.Spec.ParentRefs {
+		if handle, err := r.shouldHandleParentRef(ctx, route, ref); err != nil {
+			return nil, nil, errors.Wrapf(err, "unable to check parent ref %d", i)
+		} else if !handle {
```

Comment:
```suggestion
		}
		if !handle {
```

Reply:
I actually think the current state is better since it keeps `handle` only in scope where it needs to be.